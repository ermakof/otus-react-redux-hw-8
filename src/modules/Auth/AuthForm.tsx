import React, { FC, useContext, useEffect, useState } from 'react';
import { Formik, Field, Form } from 'formik';
import styled from '@emotion/styled';
import { waitOn, waitOff } from '@src/App/actions';
import store from '@src/store';
import {
  fakeAuthProvider,
  getUserProfileFormLocalStorage,
} from '@src/modules/Auth/fakeAuthProvider';
import { IUserProfile } from '@src/model';
import useLogin from '@src/modules/Auth/useLogin';

const Root = styled.div`
  label {
    margin-right: 5px;
  }

  button {
    margin: 5px;
  }

  input {
    margin-right: 30px;
    width: 100px;
  }

  form {
    font-size: 14px;
    margin: 3vh 0;
  }
`;

export interface IUserFormProps {
  onSubmit: (values: IAuthData) => void;
}

export type IAuthData = {
  login: string;
  password: string;
};

const AuthForm: FC = () => {
  const { dispatch } = useContext(store);
  const login = useLogin();

  const [authData, setAuthData] = useState<IAuthData>({ login: '', password: '' });

  useEffect(() => {
    const userProfile = getUserProfileFormLocalStorage();
    if (userProfile) {
      login(userProfile);
    }
  }, []);

  const handleSubmit = () => {
    dispatch(waitOn());
    fakeAuthProvider.signIn(authData, (userProfile: IUserProfile) => {
      login(userProfile);
      dispatch(waitOff());
    });
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.currentTarget;
    setAuthData({ ...authData, [id]: value });
  };

  return (
    <Root role="authForm">
      <Formik initialValues={{ ...authData }} onSubmit={handleSubmit}>
        <Form>
          <label htmlFor="login">Имя</label>
          <Field
            role="login"
            id="login"
            name="login"
            placeholder="Введите логин"
            value={authData.login}
            onChange={handleChange}
          />

          <label htmlFor="password">Пароль</label>
          <Field
            role="password"
            id="password"
            name="password"
            placeholder="Введите пароль"
            value={authData.password}
            onChange={handleChange}
          />

          <button role="buttonLogin" disabled={!authData.login || !authData.password} type="submit">
            Войти
          </button>
        </Form>
      </Formik>
    </Root>
  );
};

export default AuthForm;
