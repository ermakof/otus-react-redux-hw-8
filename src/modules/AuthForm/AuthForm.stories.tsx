import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs } from '@storybook/addon-knobs';

import UserForm from '@src/modules/UserForm';

export default {
  component: UserForm,
  decorators: [withKnobs],
  title: 'UserForm/UserForm',
} as ComponentMeta<typeof UserForm>;

const Template: ComponentStory<typeof UserForm> = (args) => <UserForm {...args} />;

export const Static = Template.bind({});

Static.args = {
  onSubmit: action('submitted'),
};
