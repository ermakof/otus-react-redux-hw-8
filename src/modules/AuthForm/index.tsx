import React, { FC, useEffect, useState } from 'react';
import { Formik, Field, Form } from 'formik';
import styled from '@emotion/styled';

const Root = styled.div`
  label {
    margin-right: 5px;
  }

  button {
    margin: 5px;
  }

  input {
    margin-right: 30px;
    width: 100px;
  }

  form {
    font-size: 14px;
    margin: 3vh 0;
  }
`;

export interface IUserFormProps {
  onSubmit: (values: IAuthData) => void;
}

export type IAuthData = {
  login: string;
  password: string;
};

const AuthForm: FC<IUserFormProps> = ({ onSubmit }) => {
  const [authData, setAuthData] = useState<IAuthData>({ login: '', password: '' });

  useEffect(() => {});

  const handleSubmit = () => {
    onSubmit({ ...authData });
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.currentTarget;
    setAuthData({ ...authData, [id]: value });
  };

  return (
    <Root role="authForm">
      <Formik initialValues={{ ...authData }} onSubmit={handleSubmit}>
        <Form>
          <label htmlFor="login">Имя</label>
          <Field
            role="login"
            id="login"
            name="login"
            placeholder="Введите логин"
            value={authData.login}
            onChange={handleChange}
          />

          <label htmlFor="password">Пароль</label>
          <Field
            role="password"
            id="password"
            name="password"
            placeholder="Введите пароль"
            value={authData.password}
            onChange={handleChange}
          />

          <button role="buttonLogin" disabled={!authData.login || !authData.password} type="submit">
            Войти
          </button>
        </Form>
      </Formik>
    </Root>
  );
};

export default AuthForm;
