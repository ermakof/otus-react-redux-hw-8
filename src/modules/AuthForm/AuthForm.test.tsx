import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import AuthForm from '@src/modules/AuthForm';
import userEvent from '@testing-library/user-event';

describe('AuthForm', () => {
  it('Render <AuthForm>', () => {
    const handleSubmit = jest.fn();
    const { asFragment } = render(<AuthForm onSubmit={handleSubmit} />);
    expect(asFragment()).toMatchSnapshot();
    const authForm = screen.getByRole(/authForm/gi);
    expect(authForm).toBeInTheDocument();
    const login = screen.getByRole(/^login$/gi);
    expect(login).toBeInTheDocument();
    const password = screen.getByRole(/password/gi);
    expect(password).toBeInTheDocument();
  });

  it('Change input <AuthForm>', async () => {
    const handleSubmit = jest.fn();
    render(<AuthForm onSubmit={handleSubmit} />);
    const login = screen.getByRole(/^login$/gi);
    userEvent.clear(login);
    userEvent.type(login, 'Иван');
    const password = screen.getByRole(/password/gi);
    userEvent.clear(password);
    userEvent.type(password, '123');

    userEvent.click(screen.getByRole('buttonLogin'));

    await waitFor(() =>
      expect(handleSubmit).toHaveBeenCalledWith({
        login: 'Иван',
        password: '123',
      })
    );
  });
});
