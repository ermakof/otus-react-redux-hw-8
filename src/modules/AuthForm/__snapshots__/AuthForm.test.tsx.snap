// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm Render <AuthForm> 1`] = `
<DocumentFragment>
  <div
    class="css-115ni0s"
    role="authForm"
  >
    <form
      action="#"
    >
      <label
        for="login"
      >
        Имя
      </label>
      <input
        id="login"
        name="login"
        placeholder="Введите логин"
        role="login"
        value=""
      />
      <label
        for="password"
      >
        Пароль
      </label>
      <input
        id="password"
        name="password"
        placeholder="Введите пароль"
        role="password"
        value=""
      />
      <button
        disabled=""
        role="buttonLogin"
        type="submit"
      >
        Войти
      </button>
    </form>
  </div>
</DocumentFragment>
`;
